### ISSUE TITLE
Possible race condition with chunked responses after upgrading to 3.5.0 (now with reproducer)

### ISSUE THREAD

**Issue by andre2i** [eclipse/vert.x#2292]
Hello,

after upgrading to vert.x **3.5.0** we've been seeing a consistent problem with one of our service APIs: when replying with **chunked responses over SSL**, clients are reporting encoding errors when processing the response.

The problem is that after a certain load threshold (not sure if the number of such chunks of their size but I can consistently repro it in our service with >4000 chunks) the client will receive invalid chunks and break the connection.

On the server side the problem surfaces with a long list of `ClosedChannelException` instances which for me I believe was a red herring as a root cause. Trying to debug the service, I noticed that the exceptions are legitimate and caused by a `Connection reset by peer` exception which is swallowed by netty:

```
this = {SslHandler@4076}

ctx = {DefaultChannelHandlerContext@4077} "ChannelHandlerContext(ssl, [id: 0x09b97b73, L:/127.0.0.1:8444 - R:/127.0.0.1:63298])"

cause = {IOException@4081} "java.io.IOException: Connection reset by peer"

detailMessage = "Connection reset by peer"
```

[The channel will then be closed](https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L1014) and triggering the large number of `ClosedChannelException` instances when the originally scheduled `zrange` commands complete. 

But there is no symptom on the server-side to describe an encoding problem so my hypothesis is that the client breaks the connection when it receives the first bad chunk thus triggering the `Connection reset by peer` which in turn surfaces the `ClosedChannelException`. So I think this is a symptom of a more delicate problem.

When this happens, `curl -raw` on the web server's endpoint will display some chunks but then stop with one of the following errors:

```
(...)
curl: (56) Malformed encoding found in chunked-encoding
```

```
(...)
curl: (56) Illegal or missing hexadecimal sequence in chunked-encoding
```

Other clients (see JavaScript below) report the same type of error:

```
      "name": "RequestError",
      "message": "Error: Parse Error",
      "cause": {
        "bytesParsed": 3,
        "code": "HPE_INVALID_CHUNK_SIZE",
        "cert": {}
      },
      "error": {
        "bytesParsed": 3,
        "code": "HPE_INVALID_CHUNK_SIZE",
        "cert": {}
      },
```

Also, comparing the payload observed by `curl` with what the server should provide, the chunk size appears to be correct every time but the chunk content is truncated at random points along the way.

All of the above leads me to believe that the response is indeed corrupted but I can't figure out why/when/how or by whom the chunked response is corrupted in the stack:

```
 at io.netty.handler.codec.http.HttpObjectEncoder.encodeChunkedContent(HttpObjectEncoder.java:165)
	  at io.netty.handler.codec.http.HttpObjectEncoder.encode(HttpObjectEncoder.java:138)
	  at io.vertx.core.http.impl.VertxHttpResponseEncoder.encode(VertxHttpResponseEncoder.java:35)
	  at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:88)
	  at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:738)
	  at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:730)
	  at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:816)
	  at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:723)
	  at io.netty.handler.stream.ChunkedWriteHandler.doFlush(ChunkedWriteHandler.java:304)
	  at io.netty.handler.stream.ChunkedWriteHandler.flush(ChunkedWriteHandler.java:137)
	  at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:776)
	  at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:768)
	  at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:749)
	  at io.netty.channel.ChannelDuplexHandler.flush(ChannelDuplexHandler.java:117)
	  at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:776)
	  at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:802)
	  at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:814)
	  at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:794)
	  at io.vertx.core.net.impl.ConnectionBase.write(ConnectionBase.java:100)
	  at io.vertx.core.net.impl.ConnectionBase.writeToChannel(ConnectionBase.java:110)
	  - locked <0xfd0> (a io.vertx.core.http.impl.ServerConnection)
	  at io.vertx.core.http.impl.ServerConnection.writeToChannel(ServerConnection.java:217)
	  at io.vertx.core.net.impl.ConnectionBase.writeToChannel(ConnectionBase.java:127)
	  at io.vertx.core.http.impl.HttpServerResponseImpl.write(HttpServerResponseImpl.java:611)
	  at io.vertx.core.http.impl.HttpServerResponseImpl.write(HttpServerResponseImpl.java:303)
	  at io.vertx.core.http.impl.HttpServerResponseImpl.write(HttpServerResponseImpl.java:58)
```

**Other observations:**

- the problem is isolated to versions above 3.5.0 (does not repro in 3.5.0.Beta1 or below and the service code which triggers this behavior hasn't been changed in a while) 
- it only appears with https and toggling OpenSSL has no effect
- for low number of chunks (less than 1000) it rarely reproduces on a cold service start but consistently reproduces afterwards
- the actual payload (string content) does not affect it. Considering that there are times in which it does not reproduce with the same content, I believe it behaves like a race condition and not like an input validation bug
- debugging appears to be adding synchronization and changes the behavior. I never succeeded in reproducing it while trying to observe [this breakpoint](https://github.com/netty/netty/blob/4.1/codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectEncoder.java#L195). What I did was to condition the breakpoint to fire when `contentLength == 0` (marking the end of the response). With the breakpoint enabled, it never reproduces. With the breakpoint muted, it always reproduces.
- looking at the change history, my initial idea was that the change which added this behavior was due to [this issue (Serialize messages sent to the connection in the order of the synchronization monitor)](https://github.com/eclipse/vert.x/issues/2145) which was closed with [these changes](https://github.com/eclipse/vert.x/commit/40f99b946ad499084f423bfef5eec6e8f9ebc8ef). But after confirming that all messages are written from the event-loop thread, I think it's not relevant.
- `HttpServerOptions.setMaxChunkSize` has no effect.
- `HttpServerResponse.setWriteQueueMaxSize` has no effect.

**A reproducer can be consulted [here](https://github.com/andre2i/vertx-sandbox/tree/master/chunked-reproducer).**

Any pointers/help to get to the root cause of this issue would be greatly appreciated.

**Comment by vietj** [comment]
can you try with Vert.x 3.5.1 SNAPSHOT ?

**Comment by vietj** [comment]
I will investigate your reproducer

**Comment by vietj** [comment]
I've just tried with 3.5.1-SNAPSHOT and it seems to pass, can you check on your side ?

**Comment by andre2i** [comment]
Hi Julien,
when I discovered this, I built the reproducer and then I had the service consume a vert.x core 3.5.1-SNAPSHOT jar built locally from sources (to try and figure out where the alleged bug is) and still reproduced the error.

But when bringing in all dependencies from the **Sonatype OSS Snapshot repository** it works and there is no repro.

When is the 3.5.1 release targeted?

Thanks for looking at this, @vietj
